pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,753 @@
+	}()
+
+	ps, err := c.packetSamplingLister.Get(psName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			c.cleanupPacketSampling(psName)
+			return nil
+		}
+		return err
+	}
+
+	switch ps.Status.Phase {
+	case "":
+		err = c.initPacketSampling(ps)
+	case crdv1alpha1.PacketSamplingRunning:
+		err = c.checkPacketSamplingStatus(ps)
+	default:
+		c.cleanupPacketSampling(psName)
+	}
+	return err
+
+}
+
+// Allocates a tag. If the PacketSampling request has been allocated with a tag
+// already, 0 is returned. If number of existing PacketSampling requests reaches
+// the upper limit, an error is returned.
+func (c *Controller) allocateTag(name string) (uint8, error) {
+	c.runningPacketSamplingsMutex.Lock()
+	defer c.runningPacketSamplingsMutex.Unlock()
+
+	for _, state := range c.runningPacketSamplings {
+		if state != nil && state.name == name {
+			// The packetsampling request has been processed already.
+			return 0, nil
+		}
+	}
+	for i := minTagNum; i <= maxTagNum; i += 1 {
+		if _, ok := c.runningPacketSamplings[i]; !ok {
+			c.runningPacketSamplings[i] = &packetSamplingState{
+				name: name,
+				tag:  i,
+			}
+			return i, nil
+		}
+	}
+	return 0, fmt.Errorf("number of on-going PacketSampling operations already reached the upper limit: %d", maxTagNum)
+}
+
+func (c *Controller) getUploaderByProtocol(protocol StorageProtocolType) (ftp.UpLoader, error) {
+	if protocol == sftpProtocol {
+		return c.sftpUploader, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+func (c *Controller) uploadPackets(ps *crdv1alpha1.PacketSampling, outputFile afero.File) error {
+	klog.V(2).InfoS("Uploading captured packets for PacketSampling", "name", ps.Name)
+	uploader, err := c.getUploaderByProtocol(sftpProtocol)
+	if err != nil {
+		return fmt.Errorf("failed to upload support bundle while getting uploader: %v", err)
+	}
+	serverAuth, err := ftp.ParseBundleAuth(ps.Spec.Authentication, c.kubeClient)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get authentication defined in the PacketSampling CR", "name", ps.Name, "authentication", ps.Spec.Authentication)
+		return err
+	}
+	cfg := ftp.GenSSHClientConfig(serverAuth.BasicAuthentication.Username, serverAuth.BasicAuthentication.Password)
+	fileName := c.nodeConfig.Name + "_" + string(ps.UID) + ".pcapng"
+
+	return uploader.Upload(ps.Spec.FileServer.URL, fileName, cfg, outputFile)
+
+}
+
+// initPacketSampling mark the packetsampling as running and allocate tag for it, then start the sampling. the tag will
+// serve as a unique id for concurrent processing.
+func (c *Controller) initPacketSampling(ps *crdv1alpha1.PacketSampling) error {
+	tag, err := c.allocateTag(ps.Name)
+	if err != nil {
+		return err
+	}
+	if tag == 0 {
+		return nil
+	}
+	err = c.updatePacketSamplingStatus(ps, crdv1alpha1.PacketSamplingRunning, "")
+	if err != nil {
+		c.deallocateTag(ps.Name, tag)
+	}
+	return c.startPacketSampling(ps, c.runningPacketSamplings[tag])
```

Comment:
I think this can be solved if we can force delete the tmp dir when the controller is starting. stale pcap files is meaningless if failed or agent restarted.